### PR TITLE
[lifecycle] Fix install operation

### PR DESCRIPTION
**Pull request by DelusionalOptimist**
Signed-off-by: Rudraksh Pareek <rudrakshpareek3601@gmail.com>

**Description**

This PR fixes #

**Notes for Reviewers**
* TODO: Fetch helm chart from downloaded istio tar

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.

### CODE REVIEW

**Comment by leecalcote** [comment]
Will the fetched helm template from istio tar be any different than the fetched helm template from the Istio repo?

File: istio/install.go [hunk 1]
```diff
@@ -37,18 +57,124 @@ func (istio *Istio) installIstio(del bool, version, namespace string) (string, e
...
 
-	err = istio.runIstioCtlCmd(version, del)
+	err = istio.applyHelmChart(del, version, namespace)
 	if err != nil {
-		istio.Log.Error(ErrInstallIstio(err))
-		return st, ErrInstallIstio(err)
+		return st, ErrApplyHelmChart(err)
 	}
 
+	// TODO: keep this as a fallback method
```
Comment (leecalcote) [inline istio/install.go:65]: Thank you

File: istio/install.go [hunk 2]
```diff
@@ -37,18 +57,124 @@ func (istio *Istio) installIstio(del bool, version, namespace string) (string, e
...
 		return status.Removed, nil
 	}
 	return status.Installed, nil
 }
 
+func (istio *Istio) applyHelmChart(del bool, version, namespace string) error {
+	kClient := istio.MesheryKubeclient
+
+	// charts should be here ideally
+	repo := "https://gcsweb.istio.io/gcs/istio-release/releases/charts/"
```
Comment (leecalcote) [inline istio/install.go:82]: Is this the post-release processing location for helm templates (the artifact repositories for install bundles)?

File: istio/install.go [hunk 3]
```diff
@@ -37,18 +57,124 @@ func (istio *Istio) installIstio(del bool, version, namespace string) (string, e
...
+	if !exists {
+		return "", ErrEntryWithAppVersionNotExists(chart, appVersion)
+	}
+
+	return entryMetadata.Version, nil
+}
+
+// CreateHelmIndex takes in the repo name and creates a
+// helm index for it. Helm index is basically marshalled version of
+// index.yaml file present in the remote helm repository
```
Comment (leecalcote) [inline istio/install.go:125]: Very helpful comment.

File: istio/install.go [hunk 4]
```diff
@@ -37,18 +57,124 @@ func (istio *Istio) installIstio(del bool, version, namespace string) (string, e
...
+			return v, true
+		}
+	}
+
+	return HelmEntryMetadata{}, false
+}
+
+// normalizeVerion takes in a version and adds "v" prefix
+// if it isn't already present
+func normalizeVersion(version string) string {
```
Comment (leecalcote) [inline istio/install.go:167]: The version string used is that of the release's git tag, correct?

**Review by leecalcote — APPROVED** [review]

**Comment by DelusionalOptimist** [comment]
> Will the fetched helm template from istio tar be any different than the fetched helm template from the Istio repo?

Yes it will be. They don't update the chart on their github repo https://github.com/istio/istio/blob/1.9.7/manifests/charts/base/Chart.yaml
so its still an old version. However, they have a hack in their build process that updates the version field in the produced tar.

File: istio/install.go (hunk 2 above)
Comment (DelusionalOptimist) [reply istio/install.go:82]: Yes, it is but not useful right now.
Ideally the index.yaml at this link should have metadata about all the charts but in this case its not updated with new releases.

Comment (DelusionalOptimist) [reply istio/install.go:82]: Just a placeholder for now.

File: istio/install.go (hunk 4 above)
Comment (DelusionalOptimist) [reply istio/install.go:167]: Yes, the git tag of the latest release. There can be two versions fields in helm charts
I think `AppVersion` is generally the git tag https://github.com/openservicemesh/osm/blob/7b40684d291691dac7374c70ae647889e6494b48/charts/osm/Chart.yaml#L21 
and `Version` is the version of the chart (doesn't contain a v prefix)

**Comment by utkarsh-pro** [comment]
@DelusionalOptimist @leecalcote I think at this point we are just duplicating the Traefik mesh code into each adapter. I think this should be moved to meshkit. I initially placed it in Traefik mesh adapter because that was the only adapter using it but now I think OSM and Istio are doing it too and we have 3 duplicates of the same code segment.

**Comment by leecalcote** [comment]
Yikes. 😅 Makes sense to me.

**Comment by DelusionalOptimist** [comment]
> @DelusionalOptimist @leecalcote I think at this point we are just duplicating the Traefik mesh code into each adapter. I think this should be moved to meshkit. I initially placed it in Traefik mesh adapter because that was the only adapter using it but now I think OSM and Istio are doing it too and we have 3 duplicates of the same code segment.

Yes @utkarsh-pro you're right :stuck_out_tongue:.
However both of the adapters were updated in a hurry. We wanted to get OSM using the SMI conformance action ASAP and I didn't want to waste time in getting the code into meshkit, making a release and then moving on .
For istio adapter, we wished to fix it in yesterday's bug hunt itself so I ended up doing the same over again. However I failed :`)
Anyways, I doubt this would remain the same here. Some changes might be needed and I'm looking for your thoughts.

**Comment by DelusionalOptimist** [comment]
Some cleanup left..

File: istio/install.go [hunk 5]
```diff
@@ -49,13 +87,150 @@ func (istio *Istio) installIstio(del bool, version, namespace string) (string, e
 	return status.Installed, nil
 }
 
-func (istio *Istio) runIstioCtlCmd(version string, isDel bool) error {
+func (istio *Istio) applyHelmChart(del bool, version, namespace, dirName string) error {
+	kClient := istio.MesheryKubeclient
+
+	// STUPID: apply charts one by one
```
Comment (DelusionalOptimist) [inline istio/install.go:93]: Is this fine or we need a better way way to do this?

File: istio/install.go [hunk 6]
```diff
@@ -49,13 +87,150 @@ func (istio *Istio) installIstio(del bool, version, namespace string) (string, e
...
+		Namespace:       "istio-system",
+		Delete:          del,
+		CreateNamespace: true,
+	})
+	if err != nil {
+		return err
+	}
+
+	err = kClient.ApplyHelmChart(mesherykube.ApplyHelmChartConfig{
+		LocalPath:       path.Join(downloadLocation, dirName, "manifests/charts/istio-control/istio-discovery"),
```
Comment (DelusionalOptimist) [inline istio/install.go:106]: Shall we hardcode these paths?

File: istio/install.go [hunk 7]
```diff
@@ -37,10 +49,36 @@ func (istio *Istio) installIstio(del bool, version, namespace string) (string, e
...
+	if err != nil {
+		// ErrGettingIstioRelease
+		return st, err
+	}
+
+	// Install using istioctl if explicitly stated
+	if useBin {
+		err = istio.runIstioCtlCmd(version, del, dirName)
+		if err != nil {
+			//ErrInstallUsingIstioCtl
```
Comment (DelusionalOptimist) [inline istio/install.go:63]: Shall we fallback to the other if either fails?

**Review by DelusionalOptimist — COMMENTED** [review]
@utkarsh-pro @leecalcote

File: istio/error.go [hunk 8]
```diff
@@ -215,3 +241,33 @@ func ErrIstioCoreComponentFail(err error) error {
...
+	return errors.New(ErrApplyHelmChartCode, errors.Alert, []string{"Error occured while applying Helm Chart"}, []string{err.Error()}, []string{}, []string{})
+}
+
+// ErrConvertingAppVersionToChartVersion is the error for converting app version to chart version
+func ErrConvertingAppVersionToChartVersion(err error) error {
+	return errors.New(ErrConvertingAppVersionToChartVersionCode, errors.Alert, []string{"Error occured while converting app version to chart version"}, []string{err.Error()}, []string{}, []string{})
+}
+
+// ErrCreatingHelmIndex is the error for creating helm index
+func ErrCreatingHelmIndex(err error) error {
```
Comment (utkarsh-pro) [inline istio/error.go:256]: Do we need them anymore?

File: istio/install.go (hunk 7 above)
Comment (utkarsh-pro) [reply istio/install.go:63]: @leecalcote do we have to support both istioctl and chart based installations?

File: istio/error.go [hunk 9]
```diff
@@ -8,101 +8,119 @@ import (
...
+
+	// ErrUnsupportedPlatformCode implies unavailbility of Istio on the
+	// requested platform
+	ErrUnsupportedPlatformCode = "1027"
+
+	// ErrIstioctlNotFoundCode implies istioctl couldn't be found anywhere
+	// on the fs
+	ErrIstioctlNotFoundCode = "1028"
+
+	// ErrDownladingTarCode implies error while downloading istio tar
```
Comment (DelusionalOptimist) [inline istio/error.go:115]:
```suggestion
	// ErrDownloadingTarCode implies error while downloading istio tar
```

**Comment by DelusionalOptimist** [comment]
Merging this in because all tests passed and only the brutal typo one https://github.com/meshery/meshery-istio/runs/3353998455#step:4:145 remained

**Pull request merged**